File: code/ex54.py
class Solution:
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """

        #Consider two specific conditions
        if len(matrix) == 0:
            return matrix
        elif len(matrix) == 1:
            return matrix[0]


        r = []

        #Init visited list
        visited = []
        for i in range(len(matrix)):
            l = []
            
            for j in range(len(matrix[0])):
                l.append(0)

            visited.append(l)

        
        #Define paths to 4 different directions
        def loop_x_right(self, i, j, matrix, visited):
            #If all visited, return
            if visited[i][j] == 1:
                return

            while j < len(matrix[0]):
                #If the path hit the wall, head to the next direction
                if visited[i][j] == 1:
                    loop_y_down(self, i + 1, j - 1, matrix, visited)
                else:
                    #Add path to list r
                    r.append(matrix[i][j])
                    visited[i][j] = 1

                j += 1

            loop_y_down(self, i + 1, j - 1, matrix, visited)


        def loop_y_down(self, i, j, matrix, visited):
            if visited[i][j] == 1:
                return

            while i < len(matrix):
                if visited[i][j] == 1:
                    loop_x_left(self, i - 1, j - 1, matrix, visited)
                else:
                    r.append(matrix[i][j])
                    visited[i][j] = 1

                i += 1

            loop_x_left(self, i - 1, j - 1, matrix, visited)


        def loop_x_left(self, i, j, matrix, visited):
            if visited[i][j] == 1:
                return

            while j >= 0:
                if visited[i][j] == 1:
                    loop_y_up(self, i - 1, j + 1, matrix, visited)
                else:
                    r.append(matrix[i][j])
                    visited[i][j] = 1

                j -= 1

            loop_y_up(self, i - 1, j + 1, matrix, visited)


        def loop_y_up(self, i, j, matrix, visited):
            if visited[i][j] == 1:
                return

            while i >= 0:
                if visited[i][j] == 1:
                    loop_x_right(self, i + 1, j + 1, matrix, visited)
                else:
                    r.append(matrix[i][j])
                    visited[i][j] = 1

                i -= 1

            loop_x_right(self, i, j + 1, matrix, visited)


        #The first direction
        loop_x_right(self, 0, 0, matrix, visited)

        return r


sol = Solution()
print(sol.spiralOrder([
  [1, 2, 3, 4],
  [5, 6, 7, 8],
  [9,10,11,12]
]))

File: code/ex54.cpp
class Solution {
    int dir[4][2] = {{0, 1}, {1, 0}, {-1, 0}, {0, -1}};
    int m, n;
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        m = matrix.size();    if (m == 0)    return {};
        n = matrix[0].size();    if (n == 0)    return {};
        
        vector<int> res;
        vector<vector<bool>> visited(m, vector<bool>(n));
        int i = 0, j = -1, d = 0;
        while (1) {
            int ni = i + dir[d][0], nj = j + dir[d][1];
            if (ni < 0 || nj < 0 || ni >= m || nj >= n || visited[ni][nj])    changeDir(d);
            else {
                res.push_back(matrix[ni][nj]);
                visited[ni][nj] = 1;
                i = ni;    j = nj;
            }
            if (end(visited, i, j))    break;
        }
        
        return res;
    }
    
private:
    void changeDir(int& d) {
        if (d == 3)    d = 0;
        else    d++;
    }
    
    bool end(vector<vector<bool>>& visited, int i, int j) {
        bool ifEnd = 1;
        for (auto& d : dir) {
            int ni = i + d[0], nj = j + d[1];
            if (ni < 0 || nj < 0 || ni >= m || nj >= n)    continue;
            if (!visited[ni][nj])    return 0;
        }
        return ifEnd;
    }
};